src/app/calculator/furnaces/air-heating/air-heating-form/air-heating-form.component.ts

```diff
@@ -0,0 +1,275 @@
+        H2: this.airHeatingService.roundVal(material.H2, 4),
+        C3H8: this.airHeatingService.roundVal(material.C3H8, 4),
+        C4H10_CnH2n: this.airHeatingService.roundVal(material.C4H10_CnH2n, 4),
+        H2O: this.airHeatingService.roundVal(material.H2O, 4),
+        CO: this.airHeatingService.roundVal(material.CO, 4),
+        CO2: this.airHeatingService.roundVal(material.CO2, 4),
+        SO2: this.airHeatingService.roundVal(material.SO2, 4),
+        O2: this.airHeatingService.roundVal(material.O2, 4),
+        substance: material.substance
+      });
+    } else {
+      let material = this.suiteDbService.selectSolidLiquidFlueGasMaterialById(this.form.controls.materialTypeId.value);
+      this.form.patchValue({
+        carbon: this.airHeatingService.roundVal(material.carbon, 4),
+        hydrogen: this.airHeatingService.roundVal(material.hydrogen, 4),
+        sulphur: this.airHeatingService.roundVal(material.sulphur, 4),
+        inertAsh: this.airHeatingService.roundVal(material.inertAsh, 4),
+        o2: this.airHeatingService.roundVal(material.o2, 4),
+        moisture: this.airHeatingService.roundVal(material.moisture, 4),
+        nitrogen: this.airHeatingService.roundVal(material.nitrogen, 4),
+        substance: material.substance
+      });
+    }
+    this.calculate();
+  }
+
+  focusField(str: string) {
+    if (str === 'materialTypeId' && this.inModal) {
+      str = 'materialTypeIdModal';
+    }
+    this.airHeatingService.currentField.next(str);
+  }
+
+  calculate() {
+    this.o2Warning = this.airHeatingFormService.checkO2Warning(this.form);
+    let updatedInput: AirHeatingInput 
+    if (this.form.controls.gasFuelType.value == true) {
+      updatedInput = this.airHeatingFormService.getAirHeatingInputGasMaterial(this.form);
+    } else {
+      updatedInput = this.airHeatingFormService.getAirHeatingInputSolidMaterial(this.form);
+    }
+    this.airHeatingService.airHeatingInput.next(updatedInput)
+  }
+
+  closeOperatingHoursModal() {
+    this.showOperatingHoursModal = false;
+  }
+
+  openOperatingHoursModal() {
+    this.showOperatingHoursModal = true;
+  }
+
+  updateOperatingHours(oppHours: OperatingHours) {
+    this.airHeatingService.operatingHours = oppHours;
+    this.form.controls.operatingHours.patchValue(oppHours.hoursPerYear);
+    this.calculate();
+    this.closeOperatingHoursModal();
+  }
+
+  setOpHoursModalWidth() {
+    if (this.formElement) {
+      this.formWidth = this.formElement.nativeElement.clientWidth;
+    }
+  }
+
+  showMaterialModal() {
+    this.airHeatingService.modalOpen.next(true);
+    if (this.form.controls.gasFuelType.value == true) {
+      this.gasMaterialModal.show();
+    } else {
+      this.solidMaterialModal.show();
+    }
+  }
+
+  hideMaterialModal(event?: any) {
+    if (event) {
+      this.setFuelOptions();
+      let newMaterial = this.fuelOptions.filter(material => { return material.substance === event.substance; });
```

Comment:
I know this was a copy paste, I started trying to improve this code block in various places already..

set the type => newMaterial: FlueGasMaterial | SolidLiquidFlueGasMaterial

use .find()

and then in if statement just check if defined

newMaterial.id and newMaterial.substance in the patch value.